### PR TITLE
Fix syntax error with C compiler in mxnet/c_api.h.

### PR DESCRIPTION
**Pull request by sighingnow**
Declarations and definitions in `mxnet/c_api.h` should be available for C compiler without C++ features. This patch fixes some errors about the compatibility with C compiler. Include:

1. Incomplete `MXNET_EXTERN_C` macro.
2. `<stddef.h>` should be included for `size_t`.
3. `bool` is not a valid C data type.
4. In C language, default arguments would be syntax error.

### CODE REVIEW

**Comment by piiswrong** [comment]
Currently custom ops are not tested. I'm worried this might break it. Have you tested examples/numpy-ops/custom_softmax.py?

**Comment by sighingnow** [comment]
This PR only includes some syntax changes in `include/c_api.h` for compatibility reason. I am working on Haskell interface of MXNet via CFFI and these flaws are treated as syntax error by C compiler.

I also think test cases for custom ops are necessary but this problem should be solved by another separated patch. This PR only intends to make C API header files available for pure C compiler. This patch doesn't affect the semantics of original code, as well as the structures of `struct CustomOpInfo` and `struct CustomOpPropInfo`.

**Comment by sighingnow** [comment]
Besides, I have noticed that there are active changes related to C API in the nnvm branch, is there any roadmap about merging nnvm to main branch?

**Comment by piiswrong** [comment]
My point is these changes will break customops. Have you verified that they won't?
nnvm will be merged in a few weeks. If you are sensitive to api change I recommend you use NNVM branch

**Comment by sighingnow** [comment]
examples/numpy-ops/custom_softmax.py works well after these changes (under cpu_only mode and python 3.5). Log of last epoch is as follow:

```
INFO:root:Epoch[19] Batch [100] Speed: 16290.70 samples/sec     Train-accuracy=0.993600
INFO:root:Epoch[19] Batch [200] Speed: 17254.55 samples/sec     Train-accuracy=0.993300
INFO:root:Epoch[19] Batch [300] Speed: 14289.14 samples/sec     Train-accuracy=0.994800
INFO:root:Epoch[19] Batch [400] Speed: 14979.27 samples/sec     Train-accuracy=0.994900
INFO:root:Epoch[19] Batch [500] Speed: 13735.33 samples/sec     Train-accuracy=0.993900
INFO:root:Epoch[19] Batch [600] Speed: 14553.61 samples/sec     Train-accuracy=0.993000
INFO:root:Epoch[19] Resetting Data Iterator
INFO:root:Epoch[19] Time cost=3.987
INFO:root:Epoch[19] Validation-accuracy=0.976300
```

File: include/mxnet/c_api.h [hunk 1]
```diff
@@ -128,10 +130,10 @@ struct CustomOpPropInfo {
   void* p_del;
 };
 
-typedef bool (*CustomOpPropCreator)(const char* /*op_type*/, const int /*num_kwargs*/,
-                                    const char** /*keys*/, const char** /*values*/,
-                                    CustomOpPropInfo* /*ret*/);
-}
+typedef int (*CustomOpPropCreator)(const char* /*op_type*/, const int /*num_kwargs*/,
```
Comment (piiswrong) [inline include/mxnet/c_api.h]: no extern C here?

Comment (sighingnow) [inline include/mxnet/c_api.h]: Done.

**Comment by piiswrong** [comment]
@tqchen @javelinjs @Ldpe2G Please have a look.
I think this looks fine but the interface change makes me a little nervous

**Comment by tqchen** [comment]
This looks good to me.

**Comment by piiswrong** [comment]
Should we fix the prototypes for custom ops in python side?

**Comment by tqchen** [comment]
Yes, we'd better do that. In most conventions, it won't hurt anyway, because bool is passed via integer in C ABI.

**Comment by tqchen** [comment]
As long we do not have array of bools it is fine, and the ABI should remain compatible. but a better practice would be use int instead of bool

**Comment by Ldpe2G** [comment]
Seems fine. @sighingnow have you run a scala example to make sure no breaking anything?

**Comment by sighingnow** [comment]
@Ldpe2G Sorry for the delayed response. I have run the mnist example of scala package just now and it worked well. My local environment: Ubuntu 16.04 64bit, java-8-openjdk-amd64 and scala 2.11.6 (under cpu_only mode). The end of log is as follow:

~~~
2016-11-27 20:06:11,716 [main] [ml.dmlc.mxnet.Model] [INFO] - Epoch[9] Train-accuracy=0.9935397
2016-11-27 20:06:11,717 [main] [ml.dmlc.mxnet.Model] [INFO] - Epoch[9] Time cost=2332
2016-11-27 20:06:11,870 [main] [ml.dmlc.mxnet.Model] [INFO] - Epoch[9] Validation-accuracy=0.9714543
2016-11-27 20:06:11,881 [main] [ml.dmlc.mxnet.examples.imclassification.TrainMnist] [INFO] - Finish fit ...
~~~

So, these changes don't affect original code related to FFI.

**Comment by piiswrong** [comment]
Could you also change the c_bool to c_int in python/mxnet/operator.py?
I think we can merge after that

**Comment by sighingnow** [comment]
Done. The example numpy-ops/custom_softmax.py still works well.

**Comment by sighingnow** [comment]
Ping @piiswrong Any more comment on this pull request ?

**Comment by howard0su** [comment]
bool is defined in C99.

#include <stdbool.h>

int main(void){
  bool b = false;
}

**Comment by howard0su** [comment]
We don't want to lose default value in C++. please define the following macro to keep it compatibale:
#ifdef __cplusplus
#define DEFAULT(x) = x
#else
#define DEFAULT(x)
#endif

int Func(int X DEFAULT(5))
{
	printf("%d", X);
}

**Comment by sighingnow** [comment]
Thanks very much for your suggestions @howard0su . I have make a pull request in #4476 .

**Pull request closed without merging**